feat: list group collaborators' suggestions in the access management …

mkrout opened the pull request:
…… (#442)

…- EXO-59958 - Meeds-io/MIPs#23 (#430)

Prior to this change, the groups proposed in the list of collaborators' suggestions are only the groups of which I am a member. After this change, if I am an administrator with any role, all groups are suggested and if I am a user, only the groups I am a member of.

mkrout committed:
feat: list group collaborators' suggestions in the access management … (#442)

…- EXO-59958 - Meeds-io/MIPs#23 (#430)

Prior to this change, the groups proposed in the list of collaborators'
suggestions are only the groups of which I am a member. After this
change, if I am an administrator with any role, all groups are suggested
and if I am a user, only the groups I am a member of.

Co-authored-by: IlhemEssaadi <93516534+IlhemEssaadi@users.noreply.github.com>
(cherry picked from commit 6576bce97c9a22ad358e9e1d51afda4870d3135a)

plamarque reviewed (commented):
where are the unit tests ?

mkrout committed:
feat: list group collaborators' suggestions in the access management … (#444)

…- EXO-59958 - Meeds-io/MIPs#23

<!-- Ensure to provide github issue and task id in the title -->
<!-- Choose between feat and fix in the title to differenciate a new
feature from a fix -->
<!-- Title format must be :
feat: FEATURE TITLE - MEED-XXXX - meeds-io/meeds#1234
or
fix: Fix TITLE - MEED-XXXX - meeds-io/meeds#1234
-->

<!-- Description : describe the feature/the fix by answering theses
questions : -->
<!-- Why is this change needed?-->
<!-- Prior to this change, ...-->
<!-- How does it address the issue?-->
<!-- This change ...-->

<!-- Tips :
Try To Limit Each Line to a Maximum Of 72 Characters
Provide links or keys to any relevant tickets, articles or other
resources

Remember to
- Capitalize the subject line
- Use the imperative mood in the subject line
- Do not end the subject line with a period
- Separate subject from body with a blank line
- Use the body to explain what and why vs. how
- Can use multiple lines with "-" for bullet points in body
-->

(cherry picked from commit 9eea7677df97285cd69e6958e78d036615389037)

mkrout commented:
> where are the unit tests ?

Added

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/ExtGroup.java:
```diff
@@ -29,15 +29,17 @@
  * @author <a href="mailto:boleslaw.dawidowicz at redhat.com">Boleslaw Dawidowicz</a>
  */
 public class ExtGroup extends GroupImpl implements Comparable<Object> {
-  private static final long serialVersionUID = -7379104016885124958L;
```
Useless Class change

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java:
```diff
@@ -52,6 +51,9 @@ public class GroupDAOImpl extends AbstractDAOImpl implements GroupHandler {
 
+    public static final String                                        ALL_GROUPS_TYPE                                =
+            "@@ALL_GROUPS@@";
+
```
```suggestion
```
You may need to use a single reference for this constant instead of duplicating it from `PatchedHibernateIdentityStoreImpl`

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java:
```diff
@@ -646,10 +649,89 @@ public ListAccess<Group> findGroupsByKeyword(String keyword) throws Exception {
+      IdentitySearchCriteria identitySearchCriteria = new IdentitySearchCriteriaImpl();
+      if (StringUtils.isNotBlank(keyword)) {
+        try {
+          identitySearchCriteria.nameFilter("*" + keyword + "*");
```
Do we need to sanitize the keyword to avoid SQL Injection or does `identitySearchCriteria` do the sanitization ?

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java:
```diff
@@ -646,10 +649,89 @@ public ListAccess<Group> findGroupsByKeyword(String keyword) throws Exception {
+        try {
+          identitySearchCriteria.nameFilter("*" + keyword + "*");
+        } catch (Exception e) {
+          handleException("unsupported Criteria error: ", e);
```
Without a Database interaction, we don't need to revert current transaction. Besides, when reading `IdentitySearchCriteriaImpl.nameFilter`, this can't happen, thus IMO you may ignore this catch block.

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java:
```diff
@@ -646,10 +649,89 @@ public ListAccess<Group> findGroupsByKeyword(String keyword) throws Exception {
+      }
+      if (user == null) {
+        if (log.isTraceEnabled()) {
+          Tools.logMethodOut(log, LogLevel.TRACE, "findGroupsOfUser", Collections.emptyList());
```
```suggestion
          Tools.logMethodOut(log, LogLevel.TRACE, "findGroupsOfUserByKeyword", Collections.emptyList());
```

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java:
```diff
@@ -646,10 +649,89 @@ public ListAccess<Group> findGroupsByKeyword(String keyword) throws Exception {
+                                                       String keyword,
+                                                       List<String> excludedGroupsTypes) throws Exception {
+      if (log.isTraceEnabled()) {
+        Tools.logMethodIn(log, LogLevel.TRACE, "findGroupsOfUser", new Object[] { "user", user });
```
```suggestion
          Tools.logMethodIn(log, LogLevel.TRACE, "findGroupsOfUserByKeyword", new Object[] { user, keyword, excludedGroupsTypes  });
```

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java:
```diff
@@ -646,10 +649,89 @@ public ListAccess<Group> findGroupsByKeyword(String keyword) throws Exception {
+        if (log.isTraceEnabled()) {
+          Tools.logMethodOut(log, LogLevel.TRACE, "findGroupsOfUser", Collections.emptyList());
+        }
+        return new LinkedList<>();
```
```suggestion
        return Collections.emptyList();
```

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java:
```diff
@@ -646,10 +649,89 @@ public ListAccess<Group> findGroupsByKeyword(String keyword) throws Exception {
+        allGroups =
+                  getIdentitySession().getRelationshipManager().findRelatedGroups(user, ALL_GROUPS_TYPE, identitySearchCriteria);
+      } catch (Exception e) {
+        handleException("Identity operation error: ", e);
```
This is a read method, thus no need to rollback the current transaction. Please delete this catch block.

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java:
```diff
@@ -646,10 +649,89 @@ public ListAccess<Group> findGroupsByKeyword(String keyword) throws Exception {
+            exoGroups.add(convertGroup(group));
+          }
+        } catch (Exception e) {
+          handleException("convert Group error: ", e);
```
Same as previous comment, to delete.

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/idm/GroupDAOImpl.java:
```diff
@@ -646,10 +649,89 @@ public ListAccess<Group> findGroupsByKeyword(String keyword) throws Exception {
+        }
+      }
+      if (log.isTraceEnabled()) {
+        Tools.logMethodOut(log, LogLevel.TRACE, "findAllGroups", exoGroups);
```
```suggestion
        Tools.logMethodOut(log, LogLevel.TRACE, "findGroupsOfUserByKeyword", exoGroups);
```

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/search/GroupSearchService.java:
```diff
@@ -9,4 +9,5 @@
 public interface GroupSearchService {
 
   ListAccess<Group> searchGroups(String term) throws Exception;
+
```
useless modified class

boubaker commented on component/identity/src/main/java/org/exoplatform/services/organization/search/GroupSearchServiceImpl.java:
```diff
@@ -9,7 +9,8 @@
   private OrganizationService organizationService;
-  
+
+
```
useless modified class

boubaker commented on component/identity/src/test/java/org/exoplatform/services/organization/TestOrganization.java:
```diff
@@ -210,6 +211,42 @@ public void testFindGroupOfUser() {
     }
 
+    public void testFindGroupsOfUserByKeyword() throws Exception
+    {
```
```suggestion
    public void testFindGroupsOfUserByKeyword() throws Exception {
```
formatting

boubaker commented on component/identity/src/test/java/org/exoplatform/services/organization/TestOrganization.java:
```diff
@@ -210,6 +211,42 @@ public void testFindGroupOfUser() {
+            excludedGroupsTypes.add("GTN_GROUP_TYPE");
+            assertEquals(0,groupHandler.findGroupsOfUserByKeyword("john","ad",excludedGroupsTypes).size());
+        } catch (Exception e) {
+            fail();
```
There is a throws in the method signature, thus no need to add a `try/catch` + `fail` instruction

boubaker commented on component/identity/src/test/java/org/exoplatform/services/organization/TestOrganization.java:
```diff
@@ -210,6 +211,42 @@ public void testFindGroupOfUser() {
+            assertEquals(1,groupHandler.findAllGroupsByKeyword("us",excludedGroupsTypes).size());
+            assertEquals(2,groupHandler.findAllGroupsByKeyword("test",excludedGroupsTypes).size());
+        } catch (Exception e) {
+            fail();
```
There is a throws in the method signature, thus no need to add a `try/catch` + `fail` instruction

boubaker commented on component/identity/src/test/java/org/exoplatform/services/organization/TestOrganization.java:
```diff
@@ -210,6 +211,42 @@ public void testFindGroupOfUser() {
+        GroupHandler groupHandler = organizationService.getGroupHandler();
+        List<String> excludedGroupsTypes = new ArrayList<>(List.of("GTN_ROOT_GROUP"));
+        try {
+            Collection<Group> groups = groupHandler.findAllGroupsByKeyword("us",excludedGroupsTypes);
+            assertNotNull(groups);
+            assertEquals(2,groups.size());
+            assertEquals(1,groupHandler.findAllGroupsByKeyword("ad",excludedGroupsTypes).size());
+            assertEquals(3,groupHandler.findAllGroupsByKeyword("test",excludedGroupsTypes).size());
+            excludedGroupsTypes.add("root_type");
+            assertEquals(1,groupHandler.findAllGroupsByKeyword("us",excludedGroupsTypes).size());
+            assertEquals(2,groupHandler.findAllGroupsByKeyword("test",excludedGroupsTypes).size());
```
```suggestion
            Collection<Group> groups = groupHandler.findAllGroupsByKeyword("us", excludedGroupsTypes);
            assertNotNull(groups);
            assertEquals(2, groups.size());
            assertEquals(1, groupHandler.findAllGroupsByKeyword("ad", excludedGroupsTypes).size());
            assertEquals(3, groupHandler.findAllGroupsByKeyword("test", excludedGroupsTypes).size());
            excludedGroupsTypes.add("root_type");
            assertEquals(1, groupHandler.findAllGroupsByKeyword("us", excludedGroupsTypes).size());
            assertEquals(2, groupHandler.findAllGroupsByKeyword("test", excludedGroupsTypes).size());
```

boubaker commented on component/portal/src/main/java/org/exoplatform/portal/rest/GroupRestResourcesV1.java:
```diff
@@ -42,6 +42,8 @@ public class GroupRestResourcesV1 implements ResourceContainer {
 
   public static final int     DEFAULT_OFFSET = 0;
+  public static final String     SPACES_GROUP_TYPE = ".spaces";
+  public static final String     ROOT_GROUP_TYPE = "root_type";
```
```suggestion
  public static final String     ROOT_GROUP_TYPE = "root_type";
```
Why adding the parent group as embedded exclusion ? Shouldn't we be able to search for parent groups ?

mkrout replied:
Functional decision, IMO we don't add members to these groups.

boubaker commented on component/portal/src/main/java/org/exoplatform/portal/rest/GroupRestResourcesV1.java:
```diff
@@ -42,6 +42,8 @@ public class GroupRestResourcesV1 implements ResourceContainer {
   public static final int     DEFAULT_LIMIT  = 20;
 
   public static final int     DEFAULT_OFFSET = 0;
+  public static final String     SPACES_GROUP_TYPE = ".spaces";
```
Same for spaces, IMO it's better to not embed this in REST business logic which has to be without business logic (See Software Architecture Guidelines).

mkrout replied:
I know, but we have a specific need here, and I don't want to change the organization service API on gatein and add a new method for this specific need from social, that is why I put it here.

boubaker commented on component/portal/src/main/java/org/exoplatform/portal/rest/GroupRestResourcesV1.java:
```diff
@@ -641,19 +646,25 @@ public Response getGroupsTreeMembers(@Context
-    int totalSize = UserGroupsList.size();
+    Group[] groups = null;
+    int totalSize = 0;
+    List<String> excludedGroupsTypes = new ArrayList<>(List.of(SPACES_GROUP_TYPE,ROOT_GROUP_TYPE));
```
This list has to be transmitted from REST parameters, even better, this list has to be a list of groups `excludeParentGroup=/space&excludeParentGroup=/` which has to be transformed in Service layer into PLIDM group type and not in REST Layer. In fact, the REST & Service APIs has to be independent from Storage layer that uses for now PLIDM. When we want to change the implementation of Storage layer, we will have to consider changing Service API & REST layers which shouldn't be the case.

mkrout replied:
Updated as requested

mkrout committed:
Apply suggestions from code review

Co-authored-by: Boubaker Khanfir <boubaker.khanfir@exoplatform.com>

mkrout committed:
feat: list group collaborators' suggestions in the access management - EXO-59958 - Meeds-io/MIPs#23

mkrout committed:
Merge branch 'develop' into EXO-59958

mkrout committed:
feat: list group collaborators' suggestions in the access management - EXO-59958 - Meeds-io/MIPs#23

mkrout committed:
Merge remote-tracking branch 'origin/EXO-59958' into EXO-59958

mkrout committed:
feat: list group collaborators' suggestions in the access management - EXO-59958 - Meeds-io/MIPs#23

mkrout committed:
Merge branch 'develop' into EXO-59958

mkrout committed:
feat: list group collaborators' suggestions in the access management - EXO-59958 - Meeds-io/MIPs#23

mkrout committed:
Merge remote-tracking branch 'origin/EXO-59958' into EXO-59958

mkrout committed:
feat: list group collaborators' suggestions in the access management - EXO-59958 - Meeds-io/MIPs#23

mkrout committed:
Merge branch 'develop' into EXO-59958

boubaker commented:
please delete Merge commit https://github.com/Meeds-io/gatein-portal/pull/443/commits/1b5b935b21d1a6e15d31c8879f32ec39b016dbe1 before merging

hbenali commented:
Well I guess we will start removing the `Rebase & Merge` button 👯‍♂️
